modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs

```diff
@@ -176,66 +176,183 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
-                    // NativeAOT shared library output
-                    && !File.Exists(Path.Combine(publishOutputTempDir, $"{GodotSharpDirs.ProjectAssemblyName}.{soExt}")))
+                // For iOS, we skip setting the publish output dir to a temp folder,
+                // and let the project use the default publish path for builds, because
+                // the buildInfo uses the publish dir as part of the struct identity, and BuildSystem
+                // uses that identity to validate if multiple requests are related to the same build.
+                // Since the iOS C# export is done in three steps:
+                //   publish for iOS; publish for iOS Simulator; generate the XCFramework
+                // and we want the output of all these three builds to be visible to the user as a whole,
+                // the buildInfo identity can't change throughout. Therefore, we don't set it at all.
+                if (ridOS != OS.DotNetOS.iOS)
                 {
-                    throw new NotSupportedException(
-                        "Publish succeeded but project assembly not found in the output directory");
+                    // Create temporary publish output directory
+                    publishOutputTempDir = Path.Combine(Path.GetTempPath(), "godot-publish-dotnet",
+                        $"{System.Environment.ProcessId}-{buildConfig}-{runtimeIdentifier}");
+
+                    _tempFolders.Add(publishOutputTempDir);
+
+                    if (!Directory.Exists(publishOutputTempDir))
+                        Directory.CreateDirectory(publishOutputTempDir);
                 }
 
-                var manifest = new StringBuilder();
+                // Create the buildInfo for the main build.
+                var buildInfo = BuildManager.CreatePublishBuildInfo(buildConfig, platform,
+                    ridOS, ridArch, publishOutputTempDir, includeDebugSymbols);
 
-                // Add to the exported project shared object list or packed resources.
-                foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
+                if (publishOutputTempDir == null)
                 {
-                    if (embedBuildResults)
+                    // We set this here after we create the buildInfo, so we can let the post-build
+                    // additional asset packaging steps run as normal
+                    publishOutputTempDir = buildInfo.PublishDirForMain();
+                    if (Directory.Exists(publishOutputTempDir))
                     {
-                        var filePath = SanitizeSlashes(Path.GetRelativePath(publishOutputTempDir, file));
-                        var fileData = File.ReadAllBytes(file);
-                        var hash = Convert.ToBase64String(SHA512.HashData(fileData));
+                        Directory.Delete(publishOutputTempDir, true);
```

Comment:
Deleting the `publishOutputTempDir` is probably not necessary if you add it to `_tempFolders`, then it should be removed automatically at the end of the exporting process. Then, this would be very similar to the non-iOS path so they could probably be merged.

Reply:
In this codepath, this isn't a temp dir - this is the output dir that the sdk sets by default, so I want to clear it _before_ running the build, so it doesn't have cruft from previous builds in it, but I don't want to clear it after the build is done, because the xcode project links to it and we need to keep the build artifacts around for further packaging steps.

In the non-iOS codepaths this dir is a temp dir and it's already added to the `_tempFolders`, to be cleared after the build.

Reply:
I see, I wonder if we should do the same for the other platforms instead of using `_tempFolders` then. Just so that we can avoid having two different codepaths.

I'm worried about using the default publish path though. What if the user modifies the path in their csproj? Or what if the default path changes[^1]?

[^1]: https://github.com/dotnet/designs/pull/281

Reply:
I've refactored all of it, so iOS is no longer weird, the publish path always gets used, and there's less changes to the publish codepath in general.